flink-core/src/main/java/org/apache/flink/api/connector/sink/InitContextInitializationContextAdapter.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.connector.sink;
+
+import org.apache.flink.api.common.serialization.SerializationSchema;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.util.UserCodeClassLoader;
+
+import java.util.function.Function;
+
+/**
+ * Adapter between {@link Sink.InitContext} and {@link SerializationSchema.InitializationContext}.
+ */
+public class InitContextInitializationContextAdapter
```

Comment:
I see that this is implemented similarly to `RuntimeContextSerializationInitializationContextAdapter` but has the same flaws.
In general, when implementing a small class it should not rely on a big class to be injected. That makes reuse and testing much harder. In this case, it would be much better to just pass the `UserCodeClassLoader` and the `MetricGroup` directly and have no dependency to the `InitContext`.
Now the metric group should only be added when it's actually needed, so here I would rather use a `Supplier<MetricGroup>` (it doesn't make sense that the `InitContext` is passed to this Adapter and then passed back to the `mapMetricGroup`; the `InitContext` is by definition available on call-site)

So signature should be
```
public InitContextInitializationContextAdapter(
            Supplier<MetricGroup> metricGroupSupplier, UserCodeClassLoader userCodeClassLoader)
```      
Finally we should cache the result from the Supplier most easily by using `Supplier#memoize` of Guava. I'd probably wrap the ctor parameter before assigning it to the field.

Reply:
I wonder if this should be in flink-connector-base instead.

Reply:
> I wonder if this should be in flink-connector-base instead.

It can also live in `flink-connector-base` I just refrained from putting it there because there is no sink-specific code yet and all these adapters are currently also in core.